c-ares: deprecation warnings for ares_init

```
src/event/net/cares/Channel.cxx:49:14: warning: 'ares_init' is deprecated: Use ares_init_options instead [-Wdeprecated-declarations]
  int code = ares_init(&channel);
             ^
./output/ANDROID/armeabi-v7a/lib/armv7a-linux-androideabi/include/ares.h:482:26: note: 'ares_init' has been explicitly marked deprecated here
CARES_EXTERN             CARES_DEPRECATED_FOR(ares_init_options) int ares_init(
                         ^
./output/ANDROID/armeabi-v7a/lib/armv7a-linux-androideabi/include/ares.h:153:22: note: expanded from macro 'CARES_DEPRECATED_FOR'
      __attribute__((deprecated("Use " #f " instead")))
                     ^
src/event/net/cares/Channel.cxx:62:18: warning: 'ares_getsock' is deprecated: Use ARES_OPT_EVENT_THREAD or ARES_OPT_SOCK_STATE_CB instead [-Wdeprecated-declarations]
  const auto s = ares_getsock(channel, socks, ARES_GETSOCK_MAXNUM);
                 ^
./output/ANDROID/armeabi-v7a/lib/armv7a-linux-androideabi/include/ares.h:649:14: note: 'ares_getsock' has been explicitly marked deprecated here
CARES_EXTERN CARES_DEPRECATED_FOR(
             ^
./output/ANDROID/armeabi-v7a/lib/armv7a-linux-androideabi/include/ares.h:153:22: note: expanded from macro 'CARES_DEPRECATED_FOR'
      __attribute__((deprecated("Use " #f " instead")))
                     ^
src/event/net/cares/Channel.cxx:162:5: warning: 'ares_gethostbyname' is deprecated: Use ares_getaddrinfo instead [-Wdeprecated-declarations]
    ares_gethostbyname(_channel, name, family, HostCallback, this);
    ^
./output/ANDROID/armeabi-v7a/lib/armv7a-linux-androideabi/include/ares.h:627:14: note: 'ares_gethostbyname' has been explicitly marked deprecated here
CARES_EXTERN CARES_DEPRECATED_FOR(ares_getaddrinfo) void ares_gethostbyname(
             ^
./output/ANDROID/armeabi-v7a/lib/armv7a-linux-androideabi/include/ares.h:153:22: note: expanded from macro 'CARES_DEPRECATED_FOR'
      __attribute__((deprecated("Use " #f " instead")))
                     ^
3 warnings generated.
```

--- a/src/event/net/cares/Channel.cxx
+++ b/src/event/net/cares/Channel.cxx
@@ -46,32 +46,21 @@ Channel::Channel(EventLoop &event_loop)
     : defer_update_sockets(event_loop, BIND_THIS_METHOD(UpdateSockets)),
       timeout_event(event_loop, BIND_THIS_METHOD(OnTimeout))
 {
-  int code = ares_init(&channel);
+  struct ares_options options;
+  int optmask = 0;
+  optmask |= ARES_OPT_SOCK_STATE_CB;
+
+  int code = ares_init_options(&channel,&options,optmask);
   if (code != 0) throw Error(code, "ares_init() failed");
 }
 
 Channel::~Channel() noexcept { ares_destroy(channel); }
 
+
 void
 Channel::UpdateSockets() noexcept
 {
   timeout_event.Cancel();
-  sockets.clear();
-
-  ares_socket_t socks[ARES_GETSOCK_MAXNUM];
-  const auto s = ares_getsock(channel, socks, ARES_GETSOCK_MAXNUM);
-
-  for (unsigned i = 0; i < ARES_GETSOCK_MAXNUM; ++i)
-  {
-    unsigned events = 0;
-
-    if (ARES_GETSOCK_READABLE(s, i)) events |= SocketEvent::READ;
-
-    if (ARES_GETSOCK_WRITABLE(s, i)) events |= SocketEvent::WRITE;
-
-    if (events != 0)
-      sockets.emplace_front(*this, SocketDescriptor{socks[i]}, events);
-  }
 
   struct timeval timeout_buffer;
   const auto *t = ares_timeout(channel, nullptr, &timeout_buffer);
@@ -97,8 +86,6 @@ Channel::OnSocket(SocketDescriptor fd, unsigned events) noexcept
 void
 Channel::OnTimeout() noexcept
 {
-  sockets.clear();
-
   ares_process_fd(channel, ARES_SOCKET_BAD, ARES_SOCKET_BAD);
 
   ScheduleUpdateSockets();
@@ -106,16 +93,16 @@ Channel::OnTimeout() noexcept
 
 template<typename F>
 static void
-AsSocketAddress(const struct hostent &he, const void *src, F &&f)
+AsSocketAddress(const ares_addrinfo_node* node, F &&f)
 {
-  switch (he.h_addrtype)
+  switch (node->ai_family)
   {
   case AF_INET:
   {
     struct sockaddr_in sin
     {
     };
-    memcpy(&sin.sin_addr, src, he.h_length);
+    memcpy(&sin.sin_addr, node->ai_addr,  node->ai_addrlen);
     sin.sin_family = AF_INET;
 
     f(SocketAddress((const struct sockaddr *)&sin, sizeof(sin)));
@@ -127,7 +114,7 @@ AsSocketAddress(const struct hostent &he, const void *src, F &&f)
     struct sockaddr_in6 sin
     {
     };
-    memcpy(&sin.sin6_addr, src, he.h_length);
+    memcpy(&sin.sin6_addr, node->ai_addr, node->ai_addrlen);
     sin.sin6_family = AF_INET6;
 
     f(SocketAddress((const struct sockaddr *)&sin, sizeof(sin)));
@@ -141,7 +128,6 @@ AsSocketAddress(const struct hostent &he, const void *src, F &&f)
 
 class Channel::Request final : Cancellable {
   Handler *handler;
-
   unsigned pending;
 
   bool success = false;
@@ -151,15 +137,13 @@ public:
       : handler(&_handler), pending(_pending)
   {
     assert(pending > 0);
-
     cancel_ptr = *this;
   }
 
   void Start(ares_channel _channel, const char *name, int family) noexcept
   {
-    assert(handler != nullptr);
-
-    ares_gethostbyname(_channel, name, family, HostCallback, this);
+    const struct ares_addrinfo_hints hints = { ARES_AI_CANONNAME, family, 0, 0 };
+    ares_getaddrinfo(_channel, name, NULL, &hints, HostCallback,this);
   }
 
 private:
@@ -173,18 +157,19 @@ private:
     handler = nullptr;
   }
 
-  void HostCallback(int status, struct hostent *he) noexcept;
+  void HostCallback(int status, struct ares_addrinfo *addressinfo) noexcept;
 
   static void HostCallback(void *arg, int status, int,
-                           struct hostent *hostent) noexcept
+                            struct ares_addrinfo *result) noexcept
   {
     auto &request = *(Request *)arg;
-    request.HostCallback(status, hostent);
+    request.HostCallback(status, result);
+
   }
 };
 
 inline void
-Channel::Request::HostCallback(int status, struct hostent *he) noexcept
+Channel::Request::HostCallback(int status, struct ares_addrinfo *addressinfo) noexcept
 {
   assert(pending > 0);
 
@@ -197,16 +182,16 @@ Channel::Request::HostCallback(int status, struct hostent *he) noexcept
   {
     if (status != ARES_SUCCESS)
       throw Error(status, "ares_gethostbyname() failed");
-    else if (he != nullptr)
+    else if (addressinfo != nullptr)
     {
       success = true;
-
-      for (auto i = he->h_addr_list; *i != nullptr; ++i)
-        AsSocketAddress(*he, *i,
-                        [&_handler = *handler](SocketAddress address)
-                        { _handler.OnCaresAddress(address); });
+      for (auto i = addressinfo->nodes;i != nullptr; i = i->ai_next){
+        AsSocketAddress(i,
+                         [&_handler = *handler](SocketAddress address)
+                         { _handler.OnCaresAddress(address); });
 
       if (pending == 0) handler->OnCaresSuccess();
+      }
     }
     else throw std::runtime_error("ares_gethostbyname() failed");
   }
@@ -217,9 +202,11 @@ Channel::Request::HostCallback(int status, struct hostent *he) noexcept
       if (success) handler->OnCaresSuccess();
       else handler->OnCaresError(std::current_exception());
     }
+    ares_freeaddrinfo(addressinfo);
   }
 
   if (pending == 0) delete this;
+  ares_freeaddrinfo(addressinfo);
 }
 
 void

--- a/src/event/net/cares/Channel.cxx
+++ b/src/event/net/cares/Channel.cxx
@@ -3,8 +3,8 @@
 // author: Max Kellermann <mk@cm4all.com>
 
 #include "Channel.hxx"
-#include "Handler.hxx"
 #include "Error.hxx"
+#include "Handler.hxx"
 #include "event/SocketEvent.hxx"
 #include "net/SocketAddress.hxx"
 #include "time/Convert.hxx"
@@ -16,8 +16,8 @@
 #ifdef _WIN32
 #include <winsock2.h>
 #else
-#include <netinet/in.h>
 #include <netdb.h>
+#include <netinet/in.h>
 #endif
 
 namespace Cares {
@@ -34,28 +34,65 @@ public:
     event.Schedule(events);
   }
 
+  SocketDescriptor GetSocket() const noexcept { return event.GetSocket(); }
+  void Reschedule(unsigned events) noexcept { event.Schedule(events); }
+
 private:
   void OnSocket(unsigned events) noexcept
   {
-    event.Cancel();
     channel.OnSocket(event.GetSocket(), events);
   }
 };
+
+void
+Channel::sock_state_cb(void *data, ares_socket_t socket_fd, int readable,
+                       int writable)
+{
+  auto &channel = *static_cast<Channel *>(data);
+  unsigned new_events = 0;
+  if (readable) new_events |= SocketEvent::READ;
+  if (writable) new_events |= SocketEvent::WRITE;
+
+  // Find existing watcher for this fd
+  auto prev = channel.sockets.before_begin();
+  for (auto it = channel.sockets.begin(); it != channel.sockets.end();) {
+    if (it->GetSocket().Get() == socket_fd) {
+      if (new_events == 0) {
+        // Socket closed by c-ares, remove the watcher
+        channel.sockets.erase_after(prev);
+      } else {
+        // Update existing watcher with new events
+        it->Reschedule(new_events);
+      }
+      return;
+    } else {
+      prev = it++;
+    }
+  }
+
+  // No existing watcher found, create new one if needed
+  if (new_events != 0)
+    channel.sockets.emplace_front(channel, SocketDescriptor{socket_fd},
+                                  new_events);
+}
 
 Channel::Channel(EventLoop &event_loop)
     : defer_update_sockets(event_loop, BIND_THIS_METHOD(UpdateSockets)),
       timeout_event(event_loop, BIND_THIS_METHOD(OnTimeout))
 {
   struct ares_options options;
+
+  options.sock_state_cb = &Channel::sock_state_cb;
+  options.sock_state_cb_data = this;
+
   int optmask = 0;
   optmask |= ARES_OPT_SOCK_STATE_CB;
 
-  int code = ares_init_options(&channel,&options,optmask);
-  if (code != 0) throw Error(code, "ares_init() failed");
+  int code = ares_init_options(&channel, &options, optmask);
+  if (code != 0) throw Error(code, "ares_init_options() failed");
 }
 
 Channel::~Channel() noexcept { ares_destroy(channel); }
-
 
 void
 Channel::UpdateSockets() noexcept
@@ -87,42 +124,22 @@ void
 Channel::OnTimeout() noexcept
 {
   ares_process_fd(channel, ARES_SOCKET_BAD, ARES_SOCKET_BAD);
-
   ScheduleUpdateSockets();
 }
 
-template<typename F>
+template <typename F>
 static void
-AsSocketAddress(const ares_addrinfo_node* node, F &&f)
+AsSocketAddress(const ares_addrinfo_node *node, F &&f)
 {
-  switch (node->ai_family)
-  {
+  switch (node->ai_family) {
   case AF_INET:
-  {
-    struct sockaddr_in sin
-    {
-    };
-    memcpy(&sin.sin_addr, node->ai_addr,  node->ai_addrlen);
-    sin.sin_family = AF_INET;
-
-    f(SocketAddress((const struct sockaddr *)&sin, sizeof(sin)));
-  }
-  break;
-
   case AF_INET6:
-  {
-    struct sockaddr_in6 sin
-    {
-    };
-    memcpy(&sin.sin6_addr, node->ai_addr, node->ai_addrlen);
-    sin.sin6_family = AF_INET6;
-
-    f(SocketAddress((const struct sockaddr *)&sin, sizeof(sin)));
-  }
-  break;
+    f(SocketAddress(node->ai_addr, node->ai_addrlen));
+    break;
 
   default:
-    throw std::runtime_error("Unsupported address type");
+    // Ignore unsupported families
+    break;
   }
 }
 
@@ -142,8 +159,8 @@ public:
 
   void Start(ares_channel _channel, const char *name, int family) noexcept
   {
-    const struct ares_addrinfo_hints hints = { ARES_AI_CANONNAME, family, 0, 0 };
-    ares_getaddrinfo(_channel, name, NULL, &hints, HostCallback,this);
+    const struct ares_addrinfo_hints hints = {ARES_AI_NOSORT, family, 0, 0};
+    ares_getaddrinfo(_channel, name, nullptr, &hints, HostCallback, this);
   }
 
 private:
@@ -155,58 +172,60 @@ private:
     /* c-ares doesn't support cancellation, so we emulate
        it here */
     handler = nullptr;
+    --pending;
   }
 
   void HostCallback(int status, struct ares_addrinfo *addressinfo) noexcept;
 
   static void HostCallback(void *arg, int status, int,
-                            struct ares_addrinfo *result) noexcept
+                           struct ares_addrinfo *result) noexcept
   {
     auto &request = *(Request *)arg;
     request.HostCallback(status, result);
-
   }
 };
 
 inline void
-Channel::Request::HostCallback(int status, struct ares_addrinfo *addressinfo) noexcept
+Channel::Request::HostCallback(int status,
+                               struct ares_addrinfo *addressinfo) noexcept
 {
   assert(pending > 0);
 
-  if (handler == nullptr) /* cancelled */
-    return;
-
   --pending;
 
-  try
-  {
-    if (status != ARES_SUCCESS)
-      throw Error(status, "ares_gethostbyname() failed");
-    else if (addressinfo != nullptr)
-    {
-      success = true;
-      for (auto i = addressinfo->nodes;i != nullptr; i = i->ai_next){
-        AsSocketAddress(i,
-                         [&_handler = *handler](SocketAddress address)
-                         { _handler.OnCaresAddress(address); });
-
-      if (pending == 0) handler->OnCaresSuccess();
-      }
-    }
-    else throw std::runtime_error("ares_gethostbyname() failed");
+  if (handler == nullptr) {
+    ares_freeaddrinfo(addressinfo);
+    if (pending == 0) delete this;
+    return;
   }
-  catch (...)
-  {
-    if (pending == 0)
-    {
-      if (success) handler->OnCaresSuccess();
-      else handler->OnCaresError(std::current_exception());
+
+  if (status != ARES_SUCCESS) {
+    if (!success && pending == 0) {
+      handler->OnCaresError(
+          std::make_exception_ptr(Error(status, "ares_getaddrinfo() failed")));
     }
     ares_freeaddrinfo(addressinfo);
+    if (pending == 0) delete this;
+    return;
   }
 
-  if (pending == 0) delete this;
+  if (addressinfo != nullptr) {
+    success = true;
+    for (auto node = addressinfo->nodes; node != nullptr;
+         node = node->ai_next) {
+      AsSocketAddress(node, [&_handler = *handler](SocketAddress address) {
+        _handler.OnCaresAddress(address);
+      });
+    }
+  } else {
+    if (pending == 0) {
+      handler->OnCaresError(std::make_exception_ptr(
+          std::runtime_error("ares_getaddrinfo() failed")));
+    }
+  }
+
   ares_freeaddrinfo(addressinfo);
+  if (pending == 0) delete this;
 }
 
 void

--- a/src/event/net/cares/Channel.hxx
+++ b/src/event/net/cares/Channel.hxx
@@ -32,6 +32,7 @@ class Channel {
 
   class Socket;
 
+  std::forward_list<Socket> sockets;
 
   class Request;
 
@@ -60,11 +61,11 @@ public:
 
 private:
   void UpdateSockets() noexcept;
-
   void ScheduleUpdateSockets() noexcept { defer_update_sockets.Schedule(); }
-
   void OnSocket(SocketDescriptor fd, unsigned events) noexcept;
   void OnTimeout() noexcept;
+  static void sock_state_cb(void *data, ares_socket_t socket_fd, int readable, int writable);
+
 };
 
 } // namespace Cares

--- a/src/event/net/cares/Channel.hxx
+++ b/src/event/net/cares/Channel.hxx
@@ -32,7 +32,6 @@ class Channel {
 
   class Socket;
 
-  std::forward_list<Socket> sockets;
 
   class Request;
 

